clang/docs/SYCLFERFCIntegrationHeaderFooter.md

```diff
@@ -0,0 +1,141 @@
+
+And for each kernel, each of its parameter is described in the kernel_signatures array.   The description consists of the following parts:
+
+* an encoding of the type that is captured
+* its size and its offset
+
+In the above example, the first two parameters are standard layout types (the int 'i' and the struct 'test_s'), the latter two though standard layout types are types which need special handling - accessor and sampler respectively.   The last entry acts as a terminating entry.
+
+This allows the run-time library to recreate the function object on the device.
+
+The full set of encoding is:
+
+* kind_accessor  for sycl accessor types
+* kind_std_layout for standard layout types
+* kind_sampler for sycl sampler types
+* kind_pointer for pointer types
+* kind_specialization_constants_buffer for initializing specialization constants
+* kind_stream for sycl stream types and
+* kind_invalid used to indicate the terminal value in the signatures array.
+
+
+# Integration Footer
+An integration footer is a source file produced during device compilation that is intended to be post-included during host compilation of the same source file.
+
+The SYCL 2020 Specification defines a specialization constant as a constant variable where the value is not known until compilation of the SYCL kernel function.
+
+Consider the following example:
+```
+  constexpr specialiation_id<int> int_const;
+  class Wrapper {
+  public:
+    static constexpr specialization_id<float> float_const;
+  };
+```
+
+Also note that the SYCL header files declare, but not define, a special function to obtain the numeric id of a specialization constant.
+```
+     namespace detail {
+         template<auto &SpecConstName>
+         inline const char *get_spec_constant_symbolic_ID();
+     }
+```
+
+In order to accommodate both implementations that have native support for this as well as those that do not, we propose the following integration footer mechanism.
+
+Definition of that function template is provided in the form of an integration footer file:
+```
+   namespace detail {
+     template<>
+     inline const char *get_spec_constant_symbolic_ID<int_const>() {
+       return "unique_name_for_int_const";
+     }
+     template<>
+     inline const char *get_spec_constant_symbolic_ID<Wrapper::float_const>() {
+       return "unique_name_for_Wrapper_float_const";
+     }
+   }
+```
+
+This footer file is appended at the end of the translation unit for the host compilation.
+
+All specialization constants used within a program are bundled together and stored into a single buffer, which is passed as implicit kernel argument. The layout of that buffer is well-defined and known to both the compiler and the runtime, so when user sets the value of a specialization constant, that value is being copied into particular place within that buffer and once the constant is requested in device code, the compiler generates a load from the same place of the buffer.
```

Comment:
I'm not sure if this paragraph is relevant to the integration header / footer design.  This part of the design is also specific to targets that don't have native support for specialization constants, so it doesn't make a lot of sense without an explanation of the difference between native vs. emulated support.  I wonder if it would just be cleaner to remove this paragraph.